File: sdks/go/observability/prometheus.go
package observability

import (
	"fmt"
	"math/big"
	"sort"
	"strings"
)

// ToPrometheusMetrics formats PaymentMetrics as Prometheus text exposition format.
func ToPrometheusMetrics(m PaymentMetrics) string {
	var b strings.Builder

	// Counters
	writeHelp(&b, "t402_payments_attempted_total", "Total payment attempts")
	writeType(&b, "t402_payments_attempted_total", "counter")
	fmt.Fprintf(&b, "t402_payments_attempted_total %d\n", m.TotalAttempted)

	writeHelp(&b, "t402_payments_successful_total", "Total successful payments")
	writeType(&b, "t402_payments_successful_total", "counter")
	fmt.Fprintf(&b, "t402_payments_successful_total %d\n", m.TotalSuccessful)

	writeHelp(&b, "t402_payments_failed_total", "Total failed payments")
	writeType(&b, "t402_payments_failed_total", "counter")
	fmt.Fprintf(&b, "t402_payments_failed_total %d\n", m.TotalFailed)

	// Latencies
	writeHelp(&b, "t402_verify_latency_ms", "Average verification latency in milliseconds")
	writeType(&b, "t402_verify_latency_ms", "gauge")
	fmt.Fprintf(&b, "t402_verify_latency_ms %.2f\n", m.AvgVerifyLatencyMs)

	writeHelp(&b, "t402_settle_latency_ms", "Average settlement latency in milliseconds")
	writeType(&b, "t402_settle_latency_ms", "gauge")
	fmt.Fprintf(&b, "t402_settle_latency_ms %.2f\n", m.AvgSettleLatencyMs)

	// Per-network counts
	if len(m.CountByNetwork) > 0 {
		writeHelp(&b, "t402_payments_by_network", "Payment count by network")
		writeType(&b, "t402_payments_by_network", "gauge")
		for _, network := range sortedKeys(m.CountByNetwork) {
			fmt.Fprintf(&b, "t402_payments_by_network{network=%q} %d\n", network, m.CountByNetwork[network])
		}
	}

	// Per-network amounts
	if len(m.AmountByNetwork) > 0 {
		writeHelp(&b, "t402_amount_by_network", "Total payment amount by network (smallest unit)")
		writeType(&b, "t402_amount_by_network", "gauge")
		for _, network := range sortedBigIntKeys(m.AmountByNetwork) {
			fmt.Fprintf(&b, "t402_amount_by_network{network=%q} %s\n", network, m.AmountByNetwork[network].String())
		}
	}

	// Failure reasons
	if len(m.FailureReasons) > 0 {
		writeHelp(&b, "t402_failure_reasons", "Payment failure count by reason")
		writeType(&b, "t402_failure_reasons", "gauge")
		for _, reason := range sortedKeys(m.FailureReasons) {
			fmt.Fprintf(&b, "t402_failure_reasons{reason=%q} %d\n", reason, m.FailureReasons[reason])
		}
	}

	return b.String()
}

func writeHelp(b *strings.Builder, name, help string) {
	fmt.Fprintf(b, "# HELP %s %s\n", name, help)
}

func writeType(b *strings.Builder, name, metricType string) {
	fmt.Fprintf(b, "# TYPE %s %s\n", name, metricType)
}

func sortedKeys(m map[string]int) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

func sortedBigIntKeys(m map[string]*big.Int) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

File: sdks/typescript/packages/observability/src/prometheus.ts
import type { PaymentMetrics } from "./types";

/**
 * Escape a Prometheus label value.
 * Backslashes, double quotes, and newlines must be escaped.
 */
function escapeLabelValue(value: string): string {
  return value.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n");
}

/**
 * Format a single Prometheus metric line.
 */
function formatMetric(name: string, labels: Record<string, string>, value: number | bigint): string {
  const labelParts = Object.entries(labels)
    .map(([k, v]) => `${k}="${escapeLabelValue(v)}"`)
    .join(",");

  const labelStr = labelParts ? `{${labelParts}}` : "";
  return `${name}${labelStr} ${value}`;
}

/**
 * Convert PaymentMetrics to Prometheus text exposition format.
 *
 * Generates the following metric families:
 * - `t402_payments_total` — counter by status and network
 * - `t402_payment_duration_seconds` — gauge for average phase durations
 * - `t402_payment_amount_total` — counter for total amount by network (smallest unit)
 *
 * @param metrics - The payment metrics to format
 * @returns Prometheus text format string
 */
export function toPrometheusMetrics(metrics: PaymentMetrics): string {
  const lines: string[] = [];

  // t402_payments_total
  lines.push("# HELP t402_payments_total Total number of payment flows by status and network.");
  lines.push("# TYPE t402_payments_total counter");
  lines.push(formatMetric("t402_payments_total", { status: "attempted" }, metrics.totalAttempted));
  lines.push(formatMetric("t402_payments_total", { status: "successful" }, metrics.totalSuccessful));
  lines.push(formatMetric("t402_payments_total", { status: "failed" }, metrics.totalFailed));

  // Per-network counts
  for (const [network, count] of Object.entries(metrics.countByNetwork)) {
    lines.push(formatMetric("t402_payments_total", { status: "successful", network }, count));
  }

  // t402_payment_duration_seconds
  lines.push("");
  lines.push(
    "# HELP t402_payment_duration_seconds Average payment phase duration in seconds.",
  );
  lines.push("# TYPE t402_payment_duration_seconds gauge");
  lines.push(
    formatMetric(
      "t402_payment_duration_seconds",
      { phase: "verify" },
      metrics.avgVerifyLatencyMs / 1000,
    ),
  );
  lines.push(
    formatMetric(
      "t402_payment_duration_seconds",
      { phase: "settle" },
      metrics.avgSettleLatencyMs / 1000,
    ),
  );

  // t402_payment_amount_total
  lines.push("");
  lines.push(
    "# HELP t402_payment_amount_total Total payment amount in smallest unit by network.",
  );
  lines.push("# TYPE t402_payment_amount_total counter");
  for (const [network, amount] of Object.entries(metrics.amountByNetwork)) {
    lines.push(formatMetric("t402_payment_amount_total", { network }, amount));
  }

  // Failure reasons
  if (Object.keys(metrics.failureReasons).length > 0) {
    lines.push("");
    lines.push("# HELP t402_payment_failures_total Payment failures by reason.");
    lines.push("# TYPE t402_payment_failures_total counter");
    for (const [reason, count] of Object.entries(metrics.failureReasons)) {
      lines.push(formatMetric("t402_payment_failures_total", { reason }, count));
    }
  }

  return lines.join("\n") + "\n";
}
